### PR TITLE
Initial release of nocostume map flag

### LINKED ISSUE #815: Disabled Costume's Sprite on certain map
Info from idRO, I wear headgear costumes, on certain maps, say it's TE Guild Dungeon and TE Castles itself, the costume's sprites are disappeared.

I was messing around on `src/map/clif.c::clif_maptypeproperty2`, there's fun note there.
````
		((map[bl->m].flag.noitemconsumption?1:0)<<8)| // DISABLE_COSTUMEITEM - Unknown - (Prevents wearing of costume items?)
````
The comment says **Prevents wearing of costume items** but, consumption wth that `noitemconsumption` flag? xDD

Here's my SS,
* **Outside TE Castle (te_prt_gld)**
screenfree thor gratis 1349
* **Inside TE Castle (te_prtcas01)**
screenfree thor gratis 1350
* **Inside TE Guild Dungeon (teg_dun01)**
screenfree thor gratis 1352

For Valkyrie/Prontera Castles, Costume sprite is still remains.

### PR DESCRIPTION
**Pull request by Atemo**
Initial release of `nocostume` map flag to follow official `DISABLE_COSTUMEITEM` map property.

* Added map flag `nocostume`, current confirmed maps are TE Castles and Guild Dungeons.
* Fixed #815.
* There's TODO note on `pc_set_costume_view`.

Signed-off-by: Cydh Ramdh <cydh@pservero.com>

### CODE REVIEW

**Comment by Atemo** [comment]
We need it for siege TE

**Comment by Lemongrass3110** [comment]
Please do not forget that this implementation breaks setlook as @cydh stated in his code. This issue needs to be resolved before we can merge it into master.

**Comment by icxbb-xx** [comment]
and break headgear with hat effect

File: src/map/atcommand.c [hunk 1]
```diff
@@ -8080,6 +8082,7 @@ ACMD_FUNC(mapflag) {
 	clif_displaymessage(sd->fd,"fog, fireworks, sakura, leaves, nogo, nobaseexp, nojobexp, nomobloot, nomvploot,");
 	clif_displaymessage(sd->fd,"nightenabled, restricted, nodrop, novending, loadevent, nochat, partylock, guildlock,");
 	clif_displaymessage(sd->fd,"reset, chmautojoin, nousecart, noitemconsumption, nosumstarmiracle, nolockon, notomb");
```
Comment (Lemongrass3110) [inline src/map/atcommand.c]: missing a comma here

File: src/map/pc.c [hunk 2]
```diff
@@ -12057,6 +11971,76 @@ void pc_show_questinfo_reinit(struct map_session_data *sd) {
...
+		if ((i = sd->equip_index[EQI_COSTUME_HEAD_MID]) != -1 && (id = sd->inventory_data[i])) {
+			if (!(id->equip&EQP_COSTUME_HEAD_TOP))
+				sd->status.head_mid = id->look;
+			else
+				sd->status.head_mid = 0;
+		}
+		if ((i = sd->equip_index[EQI_COSTUME_HEAD_TOP]) != -1 && (id = sd->inventory_data[i])) {
+			sd->status.head_top = id->look;
+		}
+		if ((i = sd->equip_index[EQP_COSTUME_GARMENT]) != -1 && (id = sd->inventory_data[i]))
```
Comment (Lemongrass3110) [inline src/map/pc.c]: wrong constant: should be EQI not EQP

File: src/map/pc.c [hunk 3]
```diff
@@ -12057,6 +11971,76 @@ void pc_show_questinfo_reinit(struct map_session_data *sd) {
...
+	if ((i = sd->equip_index[EQI_HEAD_MID]) != -1 && (id = sd->inventory_data[i])) {
+		if (!(id->equip&(EQP_HEAD_TOP)))
+			sd->status.head_mid = id->look;
+		else
+			sd->status.head_mid = 0;
+	}
+	if ((i = sd->equip_index[EQI_HEAD_TOP]) != -1 && (id = sd->inventory_data[i])) {
+		sd->status.head_top = id->look;
+	}
+	if ((i = sd->equip_index[EQP_GARMENT]) != -1 && (id = sd->inventory_data[i]))
```
Comment (Lemongrass3110) [inline src/map/pc.c]: wrong constant: should be EQI not EQP

**Comment by cydh** [comment]
Additional info
[ ] arug_dun01 and schg_dun01 also have this mapflag.

**Comment by cydh** [comment]
I'm updating this branch

File: src/map/pc.c [hunk 4]
```diff
@@ -4904,7 +4904,7 @@ int pc_useitem(struct map_session_data *sd,int n)
 #ifdef RENEWAL
 		if (pc_hasprogress(sd, WIP_DISABLE_SKILLITEM)) {
 			clif_msg(sd, WORK_IN_PROGRESS);
-			return 0;
+		return 0;
```
Comment (cydh) [inline src/map/pc.c]: need more tab

File: src/map/pc.c [hunk 5]
```diff
@@ -10793,7 +10711,7 @@ static bool pc_readdb_skilltree(char* fields[], int columns, int current)
 		if (skill_id > MAX_SKILL_ID || !skill_get_index(skill_id)) {
 			ShowWarning("pc_readdb_skilltree: Unable to load requirement skill %hu into job %d's tree.", skill_id, class_);
 			return false;
-		}
+	}
```
Comment (cydh) [inline src/map/pc.c]: need more tab

File: src/map/pc.c [hunk 6]
```diff
@@ -12238,7 +12156,72 @@ bool pc_job_can_entermap(enum e_job jobid, int m, int group_lv) {
...
+
+	if (head_low != sd->status.head_bottom || sd->setlook_head_bottom)
+		clif_changelook(&sd->bl, LOOK_HEAD_BOTTOM, sd->setlook_head_bottom ? sd->setlook_head_bottom : sd->status.head_bottom);
+	if (head_mid != sd->status.head_mid || sd->setlook_head_mid)
+		clif_changelook(&sd->bl, LOOK_HEAD_MID, sd->setlook_head_mid ? sd->setlook_head_mid : sd->status.head_mid);
+	if (head_top != sd->status.head_top || sd->setlook_head_top)
+		clif_changelook(&sd->bl, LOOK_HEAD_TOP, sd->setlook_head_top ? sd->setlook_head_top : sd->status.head_top);
+	if (robe != sd->status.robe || sd->setlook_robe)
+		clif_changelook(&sd->bl, LOOK_ROBE, sd->setlook_robe ? sd->setlook_robe : sd->status.robe);
+}/*==========================================
```
Comment (cydh) [inline src/map/pc.c]: need more line

File: src/map/clif.c [hunk 7]
```diff
@@ -6172,9 +6172,9 @@ void clif_maptypeproperty2(struct block_list *bl,enum send_target t) {
 		((map[bl->m].flag.nosumstarmiracle?0:1)<<10); // SUNMOONSTAR_MIRACLE - Unknown - (Guessing it blocks Star Gladiator's Miracle from activating)
 		//(1<<11); // Unused bits. 1 - 10 is 0x1 length and 11 is 0x15 length. May be used for future settings.
 
-	WBUFW(buf,0) = 0x99b;
-	WBUFW(buf,2) = 0x28; // Type - What is it asking for? MAPPROPERTY? MAPTYPE? I don't know. Do we even need it? [Rytech]
-	WBUFL(buf,4) = NotifyProperty;
+	WBUFW(buf,0)=0x99b;
+	WBUFW(buf,2)=0x28; // Type - What is it asking for? MAPPROPERTY? MAPTYPE? I don't know. Do we even need it? [Rytech]
+	WBUFL(buf,4)=NotifyProperty;
```
Comment (cydh) [inline src/map/clif.c]: I initially added the spaces there. :(

Comment (Lemongrass3110) [inline src/map/clif.c]: Sorry this was to keep the format changes out of it to keep the diff small and readable.

**Comment by Lemongrass3110** [comment]
So I tested this today and found a few open issues:

1) Hat Effects(from what I would guess, they should not be sent with this mapflag enabled & at least "clif-removed" from the player)

2) clif_efst_status_change_sub needs to block the following status changes from sending:

- SC_MOONSTAR
- SC_SUPER_STAR
- SC_STRANGELIGHTS
- SC_DECORATION_OF_MUSIC
- SC_LJOSALFAR
- SC_MERMAID_LONGING
- SC_HAT_EFFECT
- SC_FLOWERSMOKE
- SC_FSTONE
- SC_HAPPINESS_STAR
- SC_MAPLE_FALLS
- SC_TIME_ACCESSORY
- SC_MAGICAL_FEATHER

3) The faked "look" gets stored in the char server and displays the character not wearing any headgears in the character server.

4) The change look packets should only be sent if there really was any change to prevent multi rendering of headgears. At the end of pc_set_costume_view might solve this: 
```
	if( sd->setlook_head_bottom )
		sd->status.head_bottom = sd->setlook_head_bottom;
	if( sd->setlook_head_mid )
		sd->status.head_mid = sd->setlook_head_mid;
	if( sd->setlook_head_top )
		sd->status.head_top = sd->setlook_head_top;
	if( sd->setlook_robe )
		sd->status.robe = sd->setlook_robe;


	if (head_low != sd->status.head_bottom)
		clif_changelook(&sd->bl, LOOK_HEAD_BOTTOM, sd->status.head_bottom);
	if (head_mid != sd->status.head_mid)
		clif_changelook(&sd->bl, LOOK_HEAD_MID, sd->status.head_mid);
	if (head_top != sd->status.head_top)
		clif_changelook(&sd->bl, LOOK_HEAD_TOP, sd->status.head_top);
	if (robe != sd->status.robe)
		clif_changelook(&sd->bl, LOOK_ROBE, sd->status.robe);
```

**Comment by cydh** [comment]
1,2,3 are intended.
just our 'weird' behavior that block everything when player is hidden, so u come with assumption "if the clif is blocked, the effect will be blocked too".

For this pic below from idRO
1. Char-select in "normal" condition. (logged out in map without nocostume mapflag, ex. Prontera Inn)
2. Just a display with wearing 2 costumes, Costume Elder Crown and C Blossom Fluttering (has hateffect)
3. While hiding, hateffect is remain. It's also still displayed for others
4. Inside TE Castle (with nocostume mapflag), hateffect is still there.
5. Char-select screen when logged out in the castle.
nocostume

If you found differences on kRO/iRO, we'll follow it.
poke @Atemo

**Pull request merged**